### PR TITLE
updated feddyn implementation pytorch

### PR DESCRIPTION
**Pull request by GustavBaumgart**
The new implementation matches the algorithm implemented in the paper. This is no longer is similar to the pseudocode they provide.

The new implementation includes an updated communication protocol where all trainers send the dataset size for that round to the aggregator at the beginning of every round. This allows the aggregator to compute adaptive hyperparameters for each trainer individually and then send alpha_adpt along with the global model to begin training for the round.

### CODE REVIEW

File: lib/python/flame/common/constants.py [hunk 1]
```diff
@@ -54,3 +54,10 @@ class TrainerState(Enum):
     PRE_TRAIN = 'pre_train'
     DURING_TRAIN = 'during_train'
     POST_TRAIN = 'post_train'
+
+class AggregatorState(Enum):
+    """Enum class for aggregator state."""
+
+    PRE_TRAIN = 'pre_train'
+    DURING_TRAIN = 'during_train'
+    POST_TRAIN = 'post_train'
```
Comment (myungjin) [inline lib/python/flame/common/constants.py:63]: None of DURING_TRAIN, POST_TRAIN seem to be used in this PR. Also, the naming can be improved too.
To be precise, aggregator doesn't have "training"-related state.

Also, TrainerState and AggregatorState are exactly the same. If you want to keep the same set of states (PRE_TRAIN, DURING_TRAIN, POST_TRAIN), why not have only one enum class?

Maybe how about the following?
```
class TrainState(Enum):
    PRE = "pre"
    DURING = "during"
    POST = "post"
```

File: lib/python/flame/examples/medmnist_feddyn/aggregator/template.json [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            "name": "param-channel",
+            "pair": [
+                "trainer",
+                "aggregator"
+            ],
+            "funcTags": {
+                "aggregator": [
+                    "distribute",
+                    "aggregate",
+                    "alpha"
```
Comment (myungjin) [inline lib/python/flame/examples/medmnist_feddyn/aggregator/template.json:35]: This function tag is confusing. You may want to use `verb` to indicate it's a function. alpha sounds like a variable / argument; from optimizer alpha is indeed an argument.

File: lib/python/flame/examples/medmnist_feddyn/aggregator/template.json [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                ],
+                "trainer": [
+                    "fetch",
+                    "upload",
+                    "dataset"
```
Comment (myungjin) [inline lib/python/flame/examples/medmnist_feddyn/aggregator/template.json:40]: same here. please rename it with a verb.

File: lib/python/flame/examples/medmnist_feddyn/aggregator/template.json [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    "registry": {
+        "sort": "dummy",
+        "uri": ""
+    },
+    "selector": {
+        "sort": "default",
+        "kwargs": {}
+    },
+    "optimizer": {
+    "sort": "feddyn",
```
Comment (myungjin) [inline lib/python/flame/examples/medmnist_feddyn/aggregator/template.json:72]: wrong indentation

File: lib/python/flame/mode/horizontal/feddyn/top_aggregator.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+"""FedDyn horizontal FL top level aggregator."""
+
+import logging
+
+from ....common.util import (MLFramework, get_ml_framework_in_use,
```
Comment (myungjin) [inline lib/python/flame/mode/horizontal/feddyn/top_aggregator.py:20]: Let's use absolute path instead of relative path.

File: lib/python/flame/mode/horizontal/feddyn/top_aggregator.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        super().internal_init()
+
+    def get(self, tag: str) -> None:
+        """Get data from remote role(s)."""
+        if tag == TAG_AGGREGATE:
+            self._aggregate_weights(tag)
+        elif tag == TAG_ALPHA:
+            self._save_sizes(tag)
+
+    def _save_sizes(self, tag: str) -> None:
```
Comment (myungjin) [inline lib/python/flame/mode/horizontal/feddyn/top_aggregator.py:57]: perhaps name the function as get_dataset_size() ?

File: lib/python/flame/mode/horizontal/feddyn/top_aggregator.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        logger.debug("calling _save_sizes")
+        channel = self.cm.get_by_tag(tag)
+        if not channel:
+            return
+
+        self.dataset_sizes = dict()
+        
+        # receive dataset size from all trainers
+        all_ends = channel.all_ends()
```
Comment (myungjin) [inline lib/python/flame/mode/horizontal/feddyn/top_aggregator.py:66]: Why is this needed? why can we not use channel.ends()?

**Review by myungjin — COMMENTED** [review]
left a couple of comments. some needs an offline discussion.

File: lib/python/flame/common/constants.py (hunk 1 above)
Comment (GustavBaumgart) [reply lib/python/flame/common/constants.py:63]: Yeah, I think I can use the TrainState for both trainer and aggregator updates to the regularizer.

File: lib/python/flame/examples/medmnist_feddyn/aggregator/template.json (hunk 2 above)
Comment (GustavBaumgart) [reply lib/python/flame/examples/medmnist_feddyn/aggregator/template.json:35]: I'm thinking "get_dataset_size"

File: lib/python/flame/examples/medmnist_feddyn/aggregator/template.json (hunk 3 above)
Comment (GustavBaumgart) [reply lib/python/flame/examples/medmnist_feddyn/aggregator/template.json:40]: "send_dataset_size"?

File: lib/python/flame/examples/medmnist_feddyn/aggregator/template.json (hunk 4 above)
Comment (GustavBaumgart) [reply lib/python/flame/examples/medmnist_feddyn/aggregator/template.json:72]: Got it!

File: lib/python/flame/mode/horizontal/feddyn/top_aggregator.py (hunk 7 above)
Comment (GustavBaumgart) [reply lib/python/flame/mode/horizontal/feddyn/top_aggregator.py:66]: channel.ends() only gets the selected ends for that round. The server does need information from all clients (not just those selected) to compute alpha values for the trainers that were selected.

File: lib/python/flame/examples/medmnist_feddyn/aggregator/template.json (hunk 3 above)
Comment (myungjin) [reply lib/python/flame/examples/medmnist_feddyn/aggregator/template.json:40]: How about `uploadDatasetSize`?

File: lib/python/flame/examples/medmnist_feddyn/aggregator/template.json (hunk 2 above)
Comment (myungjin) [reply lib/python/flame/examples/medmnist_feddyn/aggregator/template.json:35]: How about `getDatasetSize`

File: lib/python/flame/mode/horizontal/coord_syncfl/trainer.py [hunk 8]
```diff
@@ -82,7 +82,7 @@ def _send_weights(self, tag: str) -> None:
         channel.await_join()
 
         self._update_weights()
-        self.regularizer.save_state(TrainerState.POST_TRAIN, loc_model=self.model)
+        self.regularizer.save_state(TrainState.POST, loc_model=self.model)
```
Comment (myungjin) [inline lib/python/flame/mode/horizontal/coord_syncfl/trainer.py:85]: why is this needed at coord_syncfl? coord_syncfl is a hierarchical fl. So, I guess we can't use feddyn with coord_syncfl.

File: lib/python/flame/mode/horizontal/feddyn/top_aggregator.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+import logging
+
+from flame.common.util import (MLFramework, get_ml_framework_in_use,
+                            weights_to_device)
+from flame.common.constants import (DeviceType, TrainState)
+from ...composer import Composer
+from ...message import MessageType
+from ...tasklet import Loop, Tasklet
+
+from ..top_aggregator import TopAggregator as BaseTopAggregator
```
Comment (myungjin) [inline lib/python/flame/mode/horizontal/feddyn/top_aggregator.py:27]: Please use the absolute path instead of the relative path.

File: lib/python/flame/mode/horizontal/syncfl/trainer.py [hunk 10]
```diff
@@ -156,7 +157,7 @@ def _send_weights(self, tag: str) -> None:
         end = channel.one_end(VAL_CH_STATE_SEND)
 
         self._update_weights()
-        self.regularizer.save_state(TrainerState.POST_TRAIN, loc_model=self.model)
+        self.regularizer.save_state(TrainState.POST, loc_model=self.model)
```
Comment (myungjin) [inline lib/python/flame/mode/horizontal/syncfl/trainer.py:160]: Now that there is a dedicated trainer.py for feddyn, do we still need this here?

**Review by myungjin — COMMENTED** [review]
left a few comments

File: lib/python/flame/mode/horizontal/coord_syncfl/trainer.py (hunk 8 above)
Comment (GustavBaumgart) [reply lib/python/flame/mode/horizontal/coord_syncfl/trainer.py:85]: At the time I just saw it was already there, so I just modified the line. Does the new PR remove it? I could also just remove it as well.

Comment (GustavBaumgart) [reply lib/python/flame/mode/horizontal/coord_syncfl/trainer.py:85]: This might have happened because this was originally (and still is) in syncfl, even though it is really only used in FedDyn as of now.

File: lib/python/flame/mode/horizontal/feddyn/top_aggregator.py (hunk 9 above)
Comment (GustavBaumgart) [reply lib/python/flame/mode/horizontal/feddyn/top_aggregator.py:27]: Oh I see, will do this for everything in lib/python/flame/mode/horizontal/feddyn

File: lib/python/flame/mode/horizontal/syncfl/trainer.py (hunk 10 above)
Comment (GustavBaumgart) [reply lib/python/flame/mode/horizontal/syncfl/trainer.py:160]: I could move this to the FedDyn trainer only (making the file longer) since it is the only place I use it now. Previously, I thought I might leave it there in case it's used in another optimizer as well, but it might not be worth it since, if removed here, save_state can me removed from other regularizers as well.

File: lib/python/flame/mode/horizontal/coord_syncfl/trainer.py (hunk 8 above)
Comment (GustavBaumgart) [reply lib/python/flame/mode/horizontal/coord_syncfl/trainer.py:85]: Decided to remove this ^

**Review by myungjin — APPROVED** [review]
lgtm

File: lib/python/flame/mode/horizontal/feddyn/top_aggregator.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        with Composer() as composer:
+            self.composer = composer
+
+            task_internal_init = Tasklet("", self.internal_init)
+
+            task_init = Tasklet("", self.initialize)
+
+            task_load_data = Tasklet("", self.load_data)
+
+            task_get_dataset = Tasklet("", self.get, TAG_GET_DATATSET_SIZE)
```
Comment (jaemin-shin) [inline lib/python/flame/mode/horizontal/feddyn/top_aggregator.py:120]: maybe we should name it as "task_get_dataset_size" instead of "task_get_dataset" as it could be misleading?

**Review by jaemin-shin — COMMENTED** [review]
Overall LGTM!

Some thoughts on sharing data sizes from trainers:
- Could sharing dataset sizes of individual trainers to aggregators be regarded as potential privacy leak? If so, what FedDyn requires for setting up its appropriate alpha value incurs tradeoff between accuracy and privacy, which we may want to consider when further investigating with FedDyn.
- How should dataset sizes of trainers be shared when there are multiple aggregators or hierarchical aggregators? I think we do not need to implement it as of now in this PR, but further evaluating it in our experiments with different configurations could be tricky.
